crates/matrix-sdk-ui/src/timeline/pagination.rs

```diff
@@ -150,216 +117,3 @@ impl super::Timeline {
         self.event_cache.pagination().status()
     }
 }
-
-/// Options for pagination.
-#[derive(Clone)]
-pub struct PaginationOptions<'a> {
-    inner: PaginationOptionsInner<'a>,
-    pub(super) wait_for_token: bool,
-}
-
-impl<'a> PaginationOptions<'a> {
-    /// Do pagination requests until we receive some events, asking the server
-    /// for the given maximum number of events.
-    ///
-    /// The server may choose to return fewer events, even if the start or end
-    /// of the visible timeline is not yet reached.
-    pub fn simple_request(event_limit: u16) -> Self {
-        Self::new(PaginationOptionsInner::SingleRequest { event_limit_if_first: Some(event_limit) })
-    }
-
-    /// Continually paginate with a fixed `limit` until at least the given
-    /// amount of timeline items have been added (unless the start or end of the
-    /// visible timeline is reached).
-    ///
-    /// The `event_limit` represents the maximum number of events the server
-    /// should return in one batch. It may choose to return fewer events per
-    /// response.
-    pub fn until_num_items(event_limit: u16, items: u16) -> Self {
-        Self::new(PaginationOptionsInner::UntilNumItems { event_limit, items })
-    }
-
-    /// Paginate once with the given initial maximum number of events, then
-    /// do more requests based on the user-provided strategy
-    /// callback.
-    ///
-    /// The callback is given numbers on the events and resulting timeline
-    /// items for the last request as well as summed over all
-    /// requests in a `paginate_backwards` call, and can decide
-    /// whether to do another request (by returning
-    /// `ControlFlow::Continue(next_event_limit)`) or not (by returning
-    /// `ControlFlow::Break(())`).
-    pub fn custom(
-        initial_event_limit: u16,
-        pagination_strategy: impl Fn(PaginationOutcome) -> ControlFlow<(), u16> + Send + Sync + 'a,
-    ) -> Self {
-        Self::new(PaginationOptionsInner::Custom {
-            event_limit_if_first: Some(initial_event_limit),
-            strategy: Arc::new(pagination_strategy),
-        })
-    }
```

Comment:
My comment has been lost by Github, so here it is again.

I clearly understand why you want to drop this code. However, it addresses a useful and pragmatic feature of Element X and other app client. Being able to predictably paginate $n$ timeline event is a useful feature.

What if we keep the `Self::custom` API inside the `EventCache` and use it inside the `Timeline`?

Reply:
So for what it's worth: we chatted about this offline, and it turns out the EX apps (our main "users") aren't using those capabilities, plus these capabilities were annoying to keep working. I've taken the liberty to remove those, but we could reintroduce them later, if there's a strong use case for them.